### PR TITLE
Add Documentation to RegexMatch and keys(::RegexMatch)

### LINKED ISSUE #40347: Document retrieving capture group names from `Regex` or `RegexMatch`
The Regex docs provide indexing with Symbol, where the symbol is the name of the capture group. 

What is lacking is a way to retrieve the capture group names from either a Match or just the Regex itself (from the docs).

```julia
s = "aaa bbb ccc"
r = r"(?<a>.*) (?<b>.*) (?<c>.*)"
m = match(r, s) # I don't think there is a way to retrieve the symbols [:a, :b, :c] from this or `r`
```
One has to know about `Base.PCRE.capture_names(m.regex.regex)`.

The context in which this arose was wanting to easily construct a `DataFrame` from a `Vector{RegexMatch}` where the capture group names become the column names. Thanks to @pdeffebach for finding the `capture_names` function for me. 

This should be a simple PR to add a line to https://docs.julialang.org/en/v1/manual/strings/#Regular-Expressions mentioning it. 

I may get to it, I may not xd

### PR DESCRIPTION
**Pull request by regier21**
Fixes #40347. I went ahead and added documentation for the the RegexMatch type as well. I also updated the manual to destructure on m rather than m.captures.

Sorry about the second PR. I had whitespace issues in the old PR (#40485) and accidentally made a mess of the the commit graph trying to fix it, so I squashed those commits and opened a new PR.

### CODE REVIEW

File: base/regex.jl [hunk 1]
```diff
@@ -145,6 +145,46 @@ in a string using an `AbstractPattern`.
 """
 abstract type AbstractMatch end
 
+"""
+A type representing a single match to a `Regex` found in a string.
```
Comment (fredrikekre) [inline base/regex.jl:149]:
```suggestion
    RegexMatch

A type representing a single match to a `Regex` found in a string.
```

File: base/regex.jl [hunk 2]
```diff
@@ -145,6 +145,46 @@ in a string using an `AbstractPattern`.
...
+[`getindex`](@ref), where keys are the the names or numbers of a capture group.
+See [`keys`](@ref keys(::RegexMatch)) for more information.
+
+# Examples
+```jldoctest
+julia> m = match(r"(?<hour>\\d+):(?<minute>\\d+)(am|pm)?", "11:30 in the morning")
+RegexMatch("11:30", hour="11", minute="30", 3=nothing)
+
+julia> hr, min, ampm = m
+RegexMatch("11:30", hour="11", minute="30", 3=nothing)
```
Comment (fredrikekre) [inline base/regex.jl:176]:
```suggestion
julia> hr, min, ampm = m;
```

File: base/regex.jl [hunk 3]
```diff
@@ -153,6 +193,28 @@ struct RegexMatch <: AbstractMatch
     regex::Regex
 end
 
+"""
+    keys(m::RegexMatch) -> Vector
+
+Returns a vector of keys for all capture groups of the underlying regex.
```
Comment (fredrikekre) [inline base/regex.jl:199]:
```suggestion
Return a vector of keys for all capture groups of the underlying regex.
```

File: base/regex.jl [hunk 4]
```diff
@@ -153,6 +193,28 @@ struct RegexMatch <: AbstractMatch
...
+A key is included even if the capture group fails to match.
+That is, `idx` will be in the return value even if `m.captures[idx] == nothing`.
```
Comment (fredrikekre) [inline base/regex.jl:201]:
```suggestion
That is, `idx` will be in the return value even if `m[idx] == nothing`.
```

**Comment by regier21** [comment]
@fredrikekre thanks for the review! Just pushed new commit accepting all suggestions.

**Comment by regier21** [comment]
A test failed but I'm pretty sure it's unrelated to the PR.

**Pull request merged**